nexus/saga-recovery/src/recovery.rs

```diff
@@ -0,0 +1,705 @@
+                    // keep track of this list and take care of it on the next
+                    // pass.
+                    if !previously_maybe_done.contains(running_saga_id) {
+                        builder.saga_maybe_done(*running_saga_id)
+                    }
+                }
+
+                Some(_found_saga) => {
+                    // The saga is in the ignore set and the database list of
+                    // running sagas.  It may have been created in the lifetime
+                    // of this program or we may have recovered it previously,
+                    // but either way, we don't have to do anything else with
+                    // this one.
+                    builder.saga_recovery_not_needed(*running_saga_id);
+                }
+            }
+        }
+
+        // Whatever's left in `running_sagas_found` at this point was found in
+        // the database list of running sagas but is not in the ignore set.  We
+        // must recover it.  (It's not possible that we already did recover it
+        // because we would have added it to our ignore set.  It's not possible
+        // that it was newly started because the starter sends a message to add
+        // this to the ignore set (and waits for it to make it to the channel)
+        // before writing the database record, and we read everything off that
+        // channel and added it to the set before calling this function.
+        for (saga_id, saga) in running_sagas_found.into_iter() {
+            builder.saga_recovery_needed(saga_id, saga);
+        }
+
+        builder.build()
+    }
+
+    /// Iterate over the sagas that need to be recovered
+    pub fn sagas_needing_recovery(
+        &self,
+    ) -> impl Iterator<Item = (&SagaId, &nexus_db_model::Saga)> + '_ {
+        self.needs_recovery.iter()
+    }
+
+    /// Iterate over the sagas that were inferred to be done
+    pub fn sagas_inferred_done(&self) -> impl Iterator<Item = SagaId> + '_ {
+        self.inferred_done.iter().copied()
+    }
+
+    /// Iterate over the sagas that should be checked on the next pass to see if
+    /// they're done
+    pub fn sagas_maybe_done(&self) -> impl Iterator<Item = SagaId> + '_ {
+        self.maybe_done.iter().copied()
+    }
+
+    /// Returns how many in-progress sagas we ignored because they were already
+    /// running
+    pub fn nskipped(&self) -> usize {
+        self.skipped_running.len()
+    }
+
+    /// Returns how many previously-in-progress sagas we now believe are done
+    pub fn ninferred_done(&self) -> usize {
+        self.inferred_done.len()
+    }
+}
+
+/// Internal helper used to construct `Plan`
+struct PlanBuilder<'a> {
+    log: &'a slog::Logger,
+    needs_recovery: BTreeMap<SagaId, nexus_db_model::Saga>,
+    skipped_running: BTreeSet<SagaId>,
+    inferred_done: BTreeSet<SagaId>,
+    maybe_done: BTreeSet<SagaId>,
+}
+
+impl<'a> PlanBuilder<'a> {
```

Comment:
nit: none of these methods need to be `pub` (and probably shouldn't be) because `PlanBuilder` is a private API.